### PR TITLE
Format list like a table

### PR DESCRIPTION
**Pull request by cmbankester**
I updated the `cmd_list` function to output the available desks as a table instead of hyphen-delimiting them. See, e.g.:

**Before:**
```bash
 desk git:(master) ✗ desk list
bid-manager - desk for working with the Bid Manager application
dotnet - desk for doing work on a .NET project
elm - Desk for working with Elm projects
fooooooooooooooooooooooooooooooooooooooo - A desk with an exceedingly long name
node.js - Desk for doing Node.js things
rails - desk for doing work on rails projects
timenator - desk for doing work on a .NET project
```

**After:**
```bash
 desk git:(master) ✗ ./desk list
NAME                                        DESCRIPTION
bid-manager                                 desk for working with the Bid Manager application
dotnet                                      desk for doing work on a .NET project
elm                                         Desk for working with Elm projects
fooooooooooooooooooooooooooooooooooooooo    A desk with an exceedingly long name
node.js                                     Desk for doing Node.js things
rails                                       desk for doing work on rails projects
timenator                                   desk for doing work on a .NET project
```

Currently no way to take the terminal width into consideration when listing, but I honestly don't see that being much of an issue these days.

I have also been considering attenuating the `list` command with an option for outputting only the names of the desks (for better integration into other systems, for instance). Thoughts?

### CODE REVIEW

**Comment by cmbankester** [comment]
Hmm, I'll have to find an alternative to the `column` command for the fish shell.

**Comment by jamesob** [comment]
@cmbankester also, make sure whatever you come up with is amenable to `cut`.

**Comment by cmbankester** [comment]
@jamesob Is the purpose of that for extracting name/description pairs?

**Comment by jamesob** [comment]
I'd like to keep desk's output as machine-friendly as possible while still remaining practical for human readability.

**Comment by jamesob** [comment]
> I have also been considering attenuating the list command with an option for outputting only the names of the desks (for better integration into other systems, for instance). Thoughts?

:+1:

**Comment by jamesob** [comment]
I'd like to hear what other users think about a display change of this sort (@ofsaleem @rnewton @jonashaag @unixorn @eush77 @mblarsen @paulmolluzzo)

**Comment by jamesob** [comment]
@cmbankester note that you'll also need to take the shell completion functions into consideration, e.g. https://github.com/jamesob/desk/blob/master/shell_plugins/bash/desk#L18

**Comment by rnewton** [comment]
I like it, but I'd prefer if the standard output was default and this was toggled via a flag.

**Comment by cmbankester** [comment]
@jamesob Really good point. I'll take a look at the completion functions that reference `list`. I'll also see about adding some tests to cover the completion functions (or at least the ones that I change).

@rnewton That's a good idea. Are you thinking with an ENV variable or with a command-line flag. Personally, I'd prefer the ENV var, but I can also understand if anyone would have reservations to that idea.

**Comment by eush77** [comment]
I like this change, this looks much more readable to me and parseable, too (you split by `\s{2,}` instead of ` - `).

**Comment by jonashaag** [comment]
> I like it, but I'd prefer if the standard output was default and this was toggled via a flag.

It should be the other way round. First, humans will type `desk ls` a lot, so it should very easy to type and have nice human-readable output. Secondly, if `desk` is integrated in another program, it doesn't hurt to add `--machine-readable` to that program's source code. So -- and this is a general rule for programs that are used by humans most of the time -- make human friendly by default and machine friendly by choice. See also Git's porcelain commands and the `--porcelain` flag. Example: `git status` is human-readable. `git status --porcelain` is machine-readable.

**Comment by rnewton** [comment]
@cmbankester I was thinking a flag + alias to your preferred output. 

@jonashaag I can get behind that argument as well. I guess the biggest thing for me is the ability to toggle the type of output you get.

**Comment by cmbankester** [comment]
I went with @jonashaag's recommendation and set the default to be human readable. Let me know if I need to make any changes:
```bash
cbankester@Christians-iMac ~/P/c/desk> ./desk list
NAME    DESCRIPTION
desk    Desk for working on desk
dotnet  Desk for working on .NET projects

cbankester@Christians-iMac ~/P/c/desk> ./desk list --machine-readable
desk - Desk for working on desk
dotnet - Desk for working on .NET projects

cbankester@Christians-iMac ~/P/c/desk> ./desk list --only-names
desk
dotnet

cbankester@Christians-iMac ~/P/c/desk> ./desk list --only-descriptions
Desk for working on desk
Desk for working on .NET projects
```

I implemented the `--only-names` (/`--only-descriptions`) flag before doing the `--machine-readable` flag (to use with the completion code), so I figured I'd leave it in. Let me know if you want me to cherry pick it out.

**Comment by paulmolluzzo** [comment]
Pretty nice idea @cmbankester! Certainly easier to know what's going on. That said, I don't know that the window width won't _ever_ be an issue. I made a ridiculously long file name and description just to illustrate the point. The screenshot is using the "default" window size for my terminal. Granted, I expand it to full width just about every single time I open the window, but not everyone does, and if you use `tmux` or `screen` and have small panes you will probably see a similar issue.

Other than that edge case I'd say this is great!

Long Names

**Comment by jamesob** [comment]
@cmbankester thanks for your continued work here, especially on keeping the tests current. A few things:

- let's get rid of the header row, if you guys don't mind,
- `--machine-readable` -> `--no-tabulate`,
- let's get rid of `--only-descriptions` if no one can think of any good usecases.

We'll want to think about applying this same sort of tabulation to `cmd_current`. My preference would be to include that change in this PR so that we maintain some consistency in the UI.

**Comment by cmbankester** [comment]
I'd recommend against using `--no-tabulate` simply because we're not actually using any tabs, just spaces. Perhaps `--columnize`?

**Comment by jamesob** [comment]
@cmbankester maybe just `--no-format`?

**Comment by cmbankester** [comment]
@jamesob I believe that make sense.

**Comment by cmbankester** [comment]
Alright, everything should be in order. Here's a sample of output from `desk` (without a subcommand):

```bash
cbankester@Christians-iMac ~/P/c/desk> ./desk
desk - Desk for working on desk

  foooooooooooooo  Alias to echo the word 'foo'
  bar              Alias to echo the word 'bar'
```
and with  the `--no-format` option:

```bash
desk - Desk for working on desk

  foooooooooooooo - Alias to echo the word 'foo'
  bar - Alias to echo the word 'bar'
```

Also, my apologies for the strangely dense collection of git commits at the bottom of this PR. I rebased on your recent changes and force-pushed onto my master.

File: desk [hunk 1]
```diff
@@ -64,7 +64,7 @@ cmd_init() {
         zsh)    local SHELLRC="${HOME}/.zshrc" ;;
     esac
 
-    read -p "Where's your shell rc file? (default: ${SHELLRC}): " \
+    read -r -p "Where's your shell rc file? (default: ${SHELLRC}): " \
```
Comment (jamesob) [inline desk]:
```
  -r   
    If this option is given, backslash does not act as an escape character.
    The backslash is considered to be part of the line. In particular, a backslash-newline
    pair can not be used as a line continuation. 
```

What's the reason you're including this change here?

File: desk [hunk 2]
```diff
@@ -114,56 +114,95 @@ cmd_run() {
 }
 
 
+# Usage:        desk list [options]
+# Description:  List all desks along with a description
+# --only-names: List only the names of the desks
+# --no-format:  Use ' - ' to separate names from descriptions
```
Comment (jamesob) [inline desk]: Ah, no more `usage()` message? I kinda liked that. 

We can implement in a follow-up PR, I guess.

File: desk (hunk 1 above)
Comment (cmbankester) [inline desk]: Wow, my bad, that was not supposed to make it in there. I was playing around with the `desk init` command in the terminal and I forgot to remove that.

Thanks for spotting that. I'll take it out.

File: desk (hunk 2 above)
Comment (cmbankester) [inline desk]: Yeah, well the reason I put the `usage` function in there was primarily for code documentation, but once I actually got around to formatting the output of the `usage` cmd (I was basing it off of the way Docker's usage help is formatted), it was painfully overladen with `\t`s and `\n`s and such. So I figured pulling it out to function descriptions would both provide good in-code documentation and allow us to automatically produce some usage help commands from the descriptions as well.

File: desk [hunk 3]
```diff
@@ -114,56 +114,95 @@ cmd_run() {
...
     if [ ! -d "${DESKS}/" ]; then
-        echo "No desk dir! Run 'desk init'." 
+        echo "No desk dir! Run 'desk init'."
         exit 1
     fi
+    local SHOW_NAMES DESKEXT AUTO_ALIGN name desc len longest out
+
+    while [[ $# -gt 0 ]]; do
+        case "$1" in
+            --only-names) test "$SHOW_NAMES" != false && SHOW_DESCRIPTIONS=false && AUTO_ALIGN=false ;;
```
Comment (jamesob) [inline desk]: when would this `test` ever not return `0`?

can't this just be 
```sh
--only-names) SHOW_DESCRIPTIONS=false ;;
```
for the same effect? `SHOW_NAMES` can then be removed entirely.

File: desk [hunk 4]
```diff
@@ -114,56 +114,95 @@ cmd_run() {
...
+            out+="$name - $desc"$'\n'
+            len=${#name}
+            (( len > longest )) && longest=$len
         fi
-    done
+    done < <(find "${DESKS}/" -name "*${DESKEXT}" -print0)
+    if [[ "$AUTO_ALIGN" != false ]]; then
+        print_aligned "$out" "$longest"
+    else
+        printf "%s" "$out"
```
Comment (jamesob) [inline desk]: `echo "$out"`?

File: desk (hunk 3 above)
Comment (cmbankester) [inline desk]: Good point, I meant to take that out.

File: desk [hunk 5]
```diff
@@ -114,56 +114,95 @@ cmd_run() {
...
+    done
+    printf "%s - %s\n" "$DESK_NAME" "$(echo_description "$DESKPATH")"
 
-        [ -z "$CALLABLES" ] || echo ""
+    if [[ -n "$CALLABLES" ]]; then
 
+        longest=0
+        out=$'\n'
         for NAME in $CALLABLES; do
             # Last clause in the grep regexp accounts for fish functions.
```
Comment (jamesob) [inline desk]: Let's keep this comment right above the `local DOCLINE` line

File: desk (hunk 2 above)
Comment (jamesob) [inline desk]: :+1:

**Comment by jamesob** [comment]
@cmbankester cool; just a few minor things. After you've pushed the changes, I'll pull and test manually. Would greatly appreciate it if you could condense this down either into a single commit or three or four meaningful segments with `git rebase -i origin/master`; if you'd rather, I can do that.

**Comment by cmbankester** [comment]
@jamesob No problem, I'll squash after I make those changes.

File: desk (hunk 4 above)
Comment (cmbankester) [inline desk]: Despite the occasional verboseness, I still prefer to use `printf` instead of `echo` virtually everywhere because the output is often much more clearly described by the command itself (i.e. I find it easier to reason about). Here, using `echo` would add an extra newline since `"$out"` already contains one. 

If you'd rather, however, I will make this change.

Comment (jamesob) [inline desk]: Gotcha; that's fine as-is.

**Comment by jamesob** [comment]
@cmbankester looks great, thanks for the good work. Do you have a bitcoin wallet QR where I might be able to send a bit of gratitude?

**Comment by cmbankester** [comment]
@jamesob Thanks so much, but the usage of desk is gratitude enough! Thanks again for open-sourcing this fine piece of software!

Feel free to @-ping me anytime you want some input on other PRs or have ideas you'd like me to implement.
